Collision: Expand defer into map/reduce, support async and be backwards compatible.

sammys opened the pull request:
For #45. Will add details in here if need be.

sammys commented:
Looks like CI is failing because I bumped @typescript-eslint/eslint-plugin and @typescript-eslint/parser to version 4 (latest).

Version 2 (used in master) doesn't support no-shadow so I had to bump it for eslint to have no complaints. Might be able to bump master to version 4 without any dire consequences.

koresar commented:
Hello @PopGoesTheWza 
This is very great PR. I'd like to merge it. But before, I want to make sure it won't break (much of) your current work?

PopGoesTheWza commented:
@koresar should be good.

koresar commented:
Great! I'll test the PR on my laptop (when it's not midnight, which is now).

sammys commented:
That last commit wraps up support for methods and initializers with all tests passing. There are a few parts that could probably be written cleaner but I don't see those as blockers.

sammys commented:
I have half implemented some more tests for the newly exposed methods on the Collision stamp and some exception throwing in those methods when things aren't right. Will push those up ASAP.

PopGoesTheWza commented:
@koresar any advise on how to integrate this PR into #78 ?

PopGoesTheWza commented:
@sammys can you add the updated `package-lock.json` file to this PR? TraviCI is confused without it...

sammys commented:
@PopGoesTheWza  That explains it. I've pushed package-lock.json with the completed implementation of *Aggregates() methods on Collision stamp.

PopGoesTheWza commented on packages/collision/index.ts:
```diff
@@ -1,233 +1,648 @@
-import { assign } from '@stamp/core';
-import { isArray, isObject, isStamp } from '@stamp/is';
+/* eslint-disable @typescript-eslint/ban-types */
+/* eslint-disable prettier/prettier */
```
@sammys can you please re-enable `prettier/prettier`? Many lines are flagged as changed just because of the missing trailing semi-colon.

sammys replied:
Done in next commit.

PopGoesTheWza reviewed (changes requested):
Does this PR requires the README documentation to be updated?

sammys commented on .eslintrc.json:
```diff
@@ -112,6 +112,8 @@
     "node/no-unpublished-require": 1,
     "node/no-unsupported-features/es-syntax": 0,
     "node/no-deprecated-api": 1,
+    "no-shadow": "off",
```
I understand. That is a non-Typescript no-shadow and I brute forced it for this branch so I'd not get warnings. Will shift it to in context. Do you think it's possible to add an adjusted eslint config for Typescript files only?

PopGoesTheWza replied:
In #78 linting is transitioning to `xo`

sammys commented on packages/compose/index.ts:
```diff
@@ -177,7 +178,8 @@ const createFactory: CreateFactory = () => {
           });
-          if (returnedValue !== undefined) instance = returnedValue;
+          // eslint-disable-next-line eqeqeq
+          if (returnedValue != undefined) instance = returnedValue as object;
```
I can't remember the specific code that triggered me to change this. Will look for it if you need me to. It turned out that this test was too specific and allowed null to be assigned to instance. In the case I encountered this was undesirable. If it is desirable I'll flip it back to !==.

koresar replied:
Very desirable. It's in the Stamp specs: https://github.com/stampit-org/stamp-specification#initializer-parameters
> If the initializer returns a value other than undefined, it replaces the instance.

sammys replied:
Ok. My bad. Null return must have been triggered by the nut behind the keyboard. Reverted in next commit.

sammys commented on packages/collision/index.d.ts:
```diff
@@ -0,0 +1,55 @@
+import { ComposeProperty, Descriptor, PropertyMap, Stamp } from '@stamp/compose';
```
Whoops! Was a little careless with the `git add` glob. Removed in next commit.

koresar commented on packages/compose/index.ts:
```diff
@@ -108,7 +109,7 @@ export interface Descriptor {
  */
 export interface Initializer {
-  (this: object, options: PropertyMap, context: InitializerContext): object | void;
+  (this: object, options: PropertyMap, context: InitializerContext): object | void | Promise<object | void>;
```
This is totally against Stamp spec. Please, let's not change anything which agains Stamp specification: https://github.com/stampit-org/stamp-specification#standalone-compose-pure-function-optional

sammys replied:
Reverted in next commit.

koresar commented on packages/compose/index.ts:
```diff
@@ -77,7 +78,7 @@ export interface Stamp extends ComposableFactory {
  * @property {Object} [staticPropertyDescriptors] ES5 Property Descriptors applied to Stamps
  */
-export interface Descriptor {
+export interface Descriptor extends PropertyMap {
```
How necessary is this change?

In my mind PropertyMap does not regulate key names and values.
Whereas Descriptor is a specific interface - regulates keys and values.

I would rather not add this if possible. However, we are flexible here. Maybe I do not understand something.

sammys replied:
Another whoops. Fixed in next commit.

koresar commented on packages/compose/index.ts:
```diff
@@ -56,6 +56,7 @@ export interface Stamp extends ComposableFactory {
    * @template Obj The type of the object instance being produced by the `Stamp`. or the type of the `Stamp` being created.
    */
   compose: ComposeProperty;
+  [key: string]: any;
```
Isn't this already defined in 
> `export interface ComposeProperty extends ComposeMethod, Descriptor, PropertyMap`
?

Why would we add it here too?

sammys replied:
Another whoops, brute force workaround to something that leaked through. Apologies. Fixed in next commit.

koresar commented:
I have couple of suggestions.

1. Start this branch of this unfinished one: https://github.com/stampit-org/stamp/tree/typescript-PaRt-two
* We would have no conflicts merging the two.
* The types are more mature in there, so might not need to do as many `compose.ts` changes.
@PopGoesTheWza Are you okay with that idea?
@sammys Is this much to ask?

2. I'm giving @sammys write permissions to this repo right now. :)

koresar commented on packages/collision/index.ts:
```diff
@@ -117,115 +417,256 @@ interface ThrowIfForbiddenOrAmbiguous {
-      return this.collisionSetup(assign({}, opts, { forbidAll: true }) as CollisionSettings);
+      return this.collisionSetup(merge({}, opts, { methods: { forbidAll: true } }) as CollisionSettings);
+    },
+    hasAggregates(this: CollisionStamp, domain: string, itemName?: string) {
```
Utility stamps (like the Collision one) tend to give its metadata (like statics) unique names. Typically, the name of the stamp is added to the static method names (not mandatory though). I.e. I'd rather rename these methods:
`hasAggregates` -> `hasCollisionAggregates` or `collisionHasAggregates`.

sammys replied:
I was able to remove the PropertyMap extension with a few extra functions to retrieve things. You'll see that in the next commit. Will making a new PR for [typescript-PaRt-two](https://github.com/stampit-org/stamp/tree/typescript-PaRt-two) unnecessarily delay access of this stuff to anyone? Since it's essentially ready for master (small details like docs remain), if you decide to merge it into master I'm happy to PR a merge of it into [typescript-PaRt-two](https://github.com/stampit-org/stamp/tree/typescript-PaRt-two). Likewise, if we have to start a new PR for it to go into [typescript-PaRt-two](https://github.com/stampit-org/stamp/tree/typescript-PaRt-two) then can do.

PopGoesTheWza replied:
Let’s keep this PR as is and mature it.  Don’t worry about the other branch yet

sammys replied:
@koresar Regarding the has/get/set methods. Naming slipped my mind during implementation. Fixed in next commit.

@PopGoesTheWza Ok

sammys commented:
> Does this PR requires the README documentation to be updated?

I will be updating the README as soon as I get an all clear for the implementation.

sammys commented:
I believe that last commit covers all the change requests. Let me know if I missed something.

Are there any remaining concerns that I need to address before updating README?

PopGoesTheWza commented:
Thanks @sammys. This PR is on a good track.

Last commit fails on build. Pro tip, run a `npm install` from the root of your local repo to check before commit.

sammys commented:
There are build issues in other packages but I didn't think those are related to this PR. I'll sort it out

sammys commented:
Had to bump Typescript to >= 3.8 so dependencies can build.

Not sure if this has implications outside the package scope... switched `this?.compose` to `this` for the following two lines:
![Screen Shot 2021-04-04 at 15 12 18](https://user-images.githubusercontent.com/1011469/113502704-2a0ba900-9558-11eb-83ff-04563e3b9c0f.png)
![Screen Shot 2021-04-04 at 15 12 49](https://user-images.githubusercontent.com/1011469/113502712-3a238880-9558-11eb-9cd4-100b6c883f1b.png)

If there are implications then I will put some tests in place and fix the code.

Had to fix a repeated type declaration in shortcut
![Screen Shot 2021-04-04 at 15 14 53](https://user-images.githubusercontent.com/1011469/113502773-971f3e80-9558-11eb-8739-0e3ccf8a85a3.png)

... and fixed the privatize-plus-collision test that used the old collision settings syntax

sammys commented:
> The build error are cause from the changes in `compose/index.ts`. My advise is to revert this file to original state. Next address any type issue locally inside `collision/index.ts`. When unsure how to correctly address a type conflict, add a `// FIXME` comment or 'force cast' type with `(someVar as unknown) as SomeType`

@PopGoesTheWza Sorry, I read this after I had committed my changes already. I can still revert if you want it that way.

koresar commented on packages/shortcut/index.ts:
```diff
@@ -32,7 +32,7 @@ interface ComposeMethod {
 
 // TODO: enhance typing
-export type ComposeProperty = ComposeMethod & Descriptor;
+export type ComposeProperty = ComposeMethod & Descriptor & PropertyMap;
```
The `.compose` property is both a method and a descriptor. But not a random key-value map. Please, remove `PropertyMap` from here too.
Sorry, I noticed too late. I'm on a vacation atm. :)

sammys replied:
Will do

koresar commented on packages/required/index.ts:
```diff
@@ -13,8 +13,8 @@ interface RequiredDescriptor extends Descriptor {
-export const required: StampMethodRequired = function required(settings): Stamp {
-  const localStamp = this?.compose ? this : Required;
+export const required: StampMethodRequired = function required(this: Stamp, settings): Stamp {
+  const localStamp = this ? this : Required;
```
This is not yet covered by a unit test, unfortunately, but `this?.compose` shouldn't have been changed. It's important to have.
Alternatively, we can change the line to:
```js
const localStamp = isStamp(this) ? this : Required;
```

sammys replied:
The compile error claims that this?.compose never fails. If this should not change then there is something else that needs to change.

sammys replied:
That will probably keep the compiler happy

sammys replied:
isStamp does indeed solve the issue. Thanks!

koresar commented on packages/named/index.ts:
```diff
@@ -6,7 +6,7 @@ import compose, { Stamp } from '@stamp/compose';
   staticProperties: {
     setName(this: Stamp, name: string): Stamp {
-      return (this?.compose ? this : Named).compose({
+      return (this ? this : Named).compose({
```
I know TS kinda protects us from this mistake. But only 50% of JS devs use TS. Please, return the logic or change to this:
```js
return (isStamp(this) ? this : Named).compose({
```

sammys replied:
The isStamp angle is what I'll try or tsc will complain

sammys replied:
isStamp does indeed solve the issue. Thanks!

koresar commented on packages/compose/index.ts:
```diff
@@ -30,14 +30,14 @@ export type ComposableFactoryParams = Parameters<ComposableFactory>;
  * @returns {Stamp} A new stamp (aka composable factory function)
  */
-interface ComposeMethod {
+interface ComposeMethod extends PropertyMap {
```
The `ComposeMethod` (actually we should some day rename it to `ComposeFunction`, the word `method` is wrong here)
is also in the specification: https://github.com/stampit-org/stamp-specification#standalone-compose-pure-function-optional
It does not have any properties on it.

Please, remove `PropertyMap` from both `ComposeMethod` here and `ComposeProperty` below.

sammys replied:
Can change it now. Will remove and report back

sammys replied:
Removed it and no other changes were necessary.

koresar commented on packages/compose/index.ts:
```diff
@@ -177,7 +177,7 @@ const createFactory: CreateFactory = () => {
             args,
           });
-          if (returnedValue !== undefined) instance = returnedValue;
+          if (returnedValue !== undefined) instance = returnedValue as object;
```
Returned values are not always objects. The specs say that it can be anything except `undefined`. Meaning: object, string, boolean, array, number, you name it.
Maybe `any` or `unknown` is better here?

sammys replied:
I was under the impression that it had to be object. Will try unknown first.

sammys replied:
See comment in next review for explanation.

koresar commented on packages/compose/index.ts:
```diff
@@ -281,13 +281,13 @@ const mergeComposable: MergeComposable = (dstDescriptor, srcComposable) => {
  * TODO
  */
-const compose: ComposeMethod = function compose(this: unknown, ...args) {
+const compose: ComposeMethod = function compose(this: Stamp | unknown, ...args: (Composable | undefined)[]): Stamp {
```
If we can remove `undefined` from here - it would be great.
If it's necessary to make `tsc` happy then let's leave it.

sammys replied:
I'll see what options we have

sammys replied:
I reverted compose/index.ts as @PopGoesTheWza suggested (makes total sense if we're going to be merging around) and then realised the code in this PR was probably written and tested using Typescript 4 before being merged into the PR branch. I'd say that is the cause of all the type strictness. Using Typescript 3 in the PR branch directly makes everything less strict and will allow us to start on the nasty merge sooner hehe.

koresar commented:
I have yet told you @sammys that the current master is not what's published to NPM. 
What happened is - I accidentally merged @PopGoesTheWza 's half baked PR. Silly me.
I hoped that there will be no other work on this repo while @PopGoesTheWza is in progress. But here comes @sammys and adds a mind blowing super functionality WITH TESTS!

The current `.js` file are what's actually published to NPM.

I don't know what it means, but I thought you need to know this Sam.

I've added few more comments. Some important, but most are minor.

In the next message I'll explain what's wrong with async initialisers.

koresar commented on packages/collision/index.ts:
```diff
@@ -1,111 +1,449 @@
+            return nextResult || this;
+          });
+        });
+      }, Promise.resolve(this || this));
```
`this || this` ? :)))

sammys replied:
Is that really in there? Lol. Probably a VS Code symbol rename victim. Will check it out.

Regarding the published vs repo thing... We have to start somewhere. I'll help @PopGoesTheWza with merging it through

koresar commented:
The async initializers are pretty difficult to implement in accordance with the specification. Here is **my** code from 5 years ago before the Spec was born: https://github.com/stampit-org/stampit/pull/133/files#diff-30f7af44dd75ee293d1a3bdab338b6604a66ce8cbc1f666444d6c982f0c7eb5dR128-R177
It's from "before async/await" times.

That little code disallowed inter-stamp operability/composability. Stamps declare that you can mish-mash anything and get your object instances from it. Which become untrue.

What I see in your current implementation:
* All initialisers will be called, even if first of them throws. - Against specs.
* Your code passes your unit tests, but won't pass more comprehensive unit testes (see link above).
* Since your code overwrites the existing `Stamp.compose.initializers` to the aggregated one that stamp won't work with a number of other stamps from this very repo. E.g. [this one](https://github.com/stampit-org/stamp/blob/de56858280411d062c567a178a0a2693ce90759b/packages/arg-over-prop/index.ts#L59), or [that one](https://github.com/stampit-org/stamp/blob/de56858280411d062c567a178a0a2693ce90759b/packages/init-property/index.ts#L25), and especially [this popular one](https://github.com/stampit-org/stamp/blob/de56858280411d062c567a178a0a2693ce90759b/packages/privatize/index.ts#L80).

Now, don't get me wrong, I really want to add async initialiser to stamps! This would be a very very very "class-beating" argument! Let's add async initiliazers in a more thought through manner we do here - discuss, update specs, update `check-compose`, release the new feature!

I have a great idea how we can extend the specification to your needs!
Add a text like this:
> If one of the intializers is detected to be an `async` one then the whole object instantiation becomes `async`.

Meaning we would detect if at least one initalizer is `async`
```js
const isAsyncfunc = f => f.constructor.name == 'AsyncFunction'
```
If yes - then return a `new Promise()` which iterates over the initalizers.

```js
    if (initializers && initializers.length > 0) {
      let returnedValue: void | object;
      const args = [options, ...moreArgs];
      for (let i = 0; i < initializers.length; i++) {
        const initializer = initializers[i];
        if (isFunction<Initializer>(initializer)) {

          // New code
          if (isAsyncFunc(initializer) return new Promise(async resolve => {
            for (; i < initializers.length; i++) {
              const initializer = initializers[i];
              if (isFunction<Initializer>(initializer)) {
                  returnedValue = await initializer.call(instance, options, { // await !!!
                  instance,
                  stamp: Stamp as Stamp,
                  args,
                });
                if (returnedValue !== undefined) instance = returnedValue;
              }
          });

          returnedValue = initializer.call(instance, options, {
            instance,
            stamp: Stamp as Stamp,
            args,
          });
          if (returnedValue !== undefined) instance = returnedValue;
        }
      }
```

WDYT?

koresar commented:
Or even better. Do not rely on semi-stable `isAsyncFunc()`. And don't rely on global `Promise` variable.

```js
    if (initializers && initializers.length > 0) {
      let returnedValue: void | object;
      const args = [options, ...moreArgs];
      for (let i = 0; i < initializers.length; i++) {
        const initializer = initializers[i];
        if (isFunction<Initializer>(initializer)) {

          // New code
          if (isAsyncFunc(initializer) return new Promise(async resolve => {
            for (; i < initializers.length; i++) {
              const initializer = initializers[i];
              if (isFunction<Initializer>(initializer)) {
                  returnedValue = await initializer.call(instance, options, { // await !!!
                  instance,
                  stamp: Stamp as Stamp,
                  args,
                });
                if (returnedValue !== undefined) instance = returnedValue;
              }
          });

          returnedValue = initializer.call(instance, options, {
            instance,
            stamp: Stamp as Stamp,
            args,
          });

          // New code
          if (returnedValue && isFunction(returnedValue.then)) return returnedValue.then(asyncReturnedValue => {
              for (i++; i < initializers.length; i++) {
                const initializer = initializers[i];
                if (isFunction<Initializer>(initializer)) {
                    asyncReturnedValue = await initializer.call(instance, options, { // await !!!
                    instance,
                    stamp: Stamp as Stamp,
                    args,
                  });
                  if (asyncReturnedValue !== undefined) instance = asyncReturnedValue;
                }
              }
              return asyncReturnedValue;
          });

          if (returnedValue !== undefined) instance = returnedValue;
        }
      }
```

koresar commented:
Also, I found the commit where we dropped Promises support. Here it is: https://github.com/stampit-org/stampit/commit/d639cb3bf72fa82452c4381f9caa65d3e16276c4#diff-30f7af44dd75ee293d1a3bdab338b6604a66ce8cbc1f666444d6c982f0c7eb5dL125
Turns out when the group of people drafted the specification (for stampit v3) they deliberately omitted thenables support.

But there is nothing that can stop us now! :)

koresar commented:
Here was another query from someone to add async initialisers support: https://github.com/stampit-org/stamp-specification/issues/120

koresar commented:
Here is an opinion on why async initialisers are bad: https://github.com/stampit-org/stamp-specification/issues/109#issuecomment-259849020

koresar commented:
Last comment for today. I've created this RFC: https://github.com/stampit-org/stamp-specification/discussions/129

sammys commented:
@koresar Thanks for digging all that up for me to look through. I'll go through it with a fresh mind. While the mind is not fresh I'll get this PR ready hehe.

sammys commented:
> What I see in your current implementation:
> 
> * All initialisers will be called, even if first of them throws. - Against specs.

Most recent commit has new tests for sync and async initialisers throwing at beginning, middle and end of the chain. I changed nothing in the `collision/index.ts` implementation. Any throw (or rejection, if the developer uses that syntax) at any point in the chain will prevent the subsequent initialisers from being executed and a promise rejection is issued to the calling code.

sammys commented:
> The async initializers were pretty difficult to implement in accordance with the specification.
> 
> What I see in your current implementation:
> 
> * All initialisers will be called, even if first of them throws. - Against specs.
> * Your code passes your unit tests, but won't pass more comprehensive unit testes (see link above).

I've hit one of these more complex scenarios in the project I'm working on and, as you expected, it does fail. In the current PR implementation the async reduce doesn't properly handle when an initialiser returns non-undefined. I've implemented a fix that _appears to_ handle it properly. While it does run properly in the project it fails tests for some very bizarre reason and that's why I haven't yet committed anything to the PR. Working on it.
